### PR TITLE
Fix supervisor healthcheck throttling

### LINKED ISSUE #37742: high cpu on dockerd/docker-containerd
Noticed high cpu (70%-80%) on dockerd and docker-containerd on master. Fresh state with no commands.

On killing containerd there seems to be a related issue where it prints hundreds of repeated error logs in a loop.

**Steps to reproduce the issue:**
1. make shell
2. ./hack/make.sh binary install-binary
3. dockerd -s overlay2 -D

Bisected to this commit:

```
dd2e19ebd58cd8896d79b4b8db61144b93717b33 is the first bad commit
commit dd2e19ebd58cd8896d79b4b8db61144b93717b33
Author: Derek McGowan <derek@mcgstyle.net>
Date:   Wed May 23 12:15:21 2018 -0700

    libcontainerd: split client and supervisor

    Adds a supervisor package for starting and monitoring containerd.
    Separates grpc connection allowing access from daemon.

    Signed-off-by: Derek McGowan <derek@mcgstyle.net>

:040000 040000 6d7fc5323fe5550f93ea429a0db7f2a19c14d5c8 901541324e0d3f2e5637dc47ac45163916e15575 M	cmd
:040000 040000 36027bc6a35d2b6234685e21e98aef4a24613055 dcb978b0422a3e07dc4d125372fd6aa6f8c76fa3 M	daemon
:040000 040000 97c89eecf0033ed4058a1d9887d461b5a0ec64a7 41cae9c84637f3da5d50c524b7e4aa1d14b2bb73 M	libcontainerd
:040000 040000 d8eb6432b3d73302ab992a1fce4331035d846630 1e41b484829e34e9cfb7890d9f18c41ffdd4ab6a M	plugin
```

@dmcgowan @andrewhsu

### PR DESCRIPTION
**Pull request by dmcgowan**
Fix default case causing the throttling to not be used.
Ensure that nil client condition is handled.

Fixes #37742

### CODE REVIEW

**Comment by kolyshkin** [comment]
Alternative: https://github.com/moby/moby/pull/37757

File: libcontainerd/supervisor/remote_daemon.go [hunk 1]
```diff
@@ -276,26 +282,28 @@ func (r *remote) monitorDaemon(ctx context.Context) {
...
 		}
 
-		tctx, cancel := context.WithTimeout(ctx, healthCheckTimeout)
-		_, err := client.IsServing(tctx)
-		cancel()
-		if err == nil {
-			if !started {
-				close(r.daemonStartCh)
-				started = true
+		if client != nil {
```
Comment (kolyshkin) [inline libcontainerd/supervisor/remote_daemon.go]: Maybe do this instead:

```go
               if r.daemonPid == -1 {
                        <.....>
                }

                if client == nil {
                        client, err = containerd.New(r.GRPC.Address, containerd.WithTimeout(60*time.Second))
                        if err != nil {
                                r.logger.WithError(err).Error("failed connecting to containerd")
                                delay = time.After(100 * time.Millisecond)
                                continue
                        }
                }

                tctx, cancel := context.WithTimeout(ctx, healthCheckTimeout)
                _, err := client.IsServing(tctx)
                cancel()
....
```

**Review by kolyshkin — CHANGES_REQUESTED** [review]

File: libcontainerd/supervisor/remote_daemon.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainerd/supervisor/remote_daemon.go]: This appears to be simpler and easier to follow/read.

Comment (dmcgowan) [reply libcontainerd/supervisor/remote_daemon.go]: I originally had a change like this, the problem I was concerned about though is if the daemon must be restarted in order to get a successful connection. There is no timeout in this flow to kill the daemon process. My thinking with the current change is that the connection logic may have already waited up to a reasonable time to connect, and if it couldn't then it should restart the whole flow, including killing containerd and starting a new process.

Comment (kolyshkin) [reply libcontainerd/supervisor/remote_daemon.go]: makes sense

**Review by kolyshkin — APPROVED** [review]

**Review by tonistiigi — APPROVED** [review]

**Review by tiborvass — APPROVED** [review]
LGTM

**Pull request merged**